lib/PolarisRoute.Class.ps1

```diff
@@ -0,0 +1,110 @@
+#
+# Copyright (c) Microsoft. All rights reserved.
+# Licensed under the MIT license. See LICENSE file in the project root for full license information.
+#
+
+
+class PolarisRoute {
+    
+    [string]
+    $Method
+
+    [scriptblock]
+    $Scriptblock
+
+    [string]
+    hidden $_Path
+    
+    [regex]
+    hidden $_Regex
+
+    PolarisRoute([string]$Method, $Path, [scriptblock]$Scriptblock) {
+        $this.Scriptblock = $Scriptblock
+        $this.Method = $Method
+        $this._Path = [PolarisRoute]::SanitizePath($Path)
+        $this._Regex = [PolarisRoute]::ConvertPathToRegex($Path)
+
+        $Path_Get = {
+            # Getter
+            return $this._Path
+        }
+
+        $Path_Set = {
+            # Setter
+            param($value)
+            $this._Path = [PolarisRoute]::SanitizePath($value)
+            $this._Regex = [PolarisRoute]::ConvertPathToRegex($value)
+        }
+
+        $PathProperty = @{
+            Name        = "Path"
+            Value       = $Path_Get
+            SecondValue = $Path_Set
+            MemberType  = "ScriptProperty"
+        }
+
+        $this | Add-Member @PathProperty
```

Comment:
Can you explain why the Add-Member stuff here is needed vs just having Path and Regex not be hidden?

Reply:
Sure let me try and talk through my thought process here and then if it seems legit we'll add in some documentation into the code to clarify things a little bit.

On the current iteration of our build we only expose the "path" that was given to Polaris. The Regular Expression that the path is converted to at run-time isn't really exposed. If someone wants to figure out exactly how the path they fed Polaris is getting converted to RegEx they have to run this method against the path that they want.

https://github.com/PowerShell/Polaris/blob/a0117afb7308273d869167af2b3d8be4f5ed8fe4/lib/Polaris.Class.ps1#L211-L232

While that doesn't sound too bad it lends itself to a bit of mystery on how things are actually working under the hood.

What I wanted was for someone to be able to run the following commands:

```pwsh
> New-PolarisRoute -Path "/hello*/:name" -Method GET -Scriptblock { $Response.Send("Hello $($Parameters.Name)") }
> Get-PolarisRoute
```
and see the following output

```
Path          Regex                  Method Scriptblock
----          -----                  ------ -----------
/hello*/:name ^/hello.*/(?<name>.+)$ GET     $Response.Send("Hello $($Parameters.Name)")
```

Converting to RegEx when the path is set also should give a bit of a performance boost as the regular expression isn't being recalculated every time the path is run as we have here:

https://github.com/PowerShell/Polaris/blob/a0117afb7308273d869167af2b3d8be4f5ed8fe4/lib/Polaris.Class.ps1#L70

So essentially I had two requirements:

1. RegEx should be a read-only field
2. RegEx should be re-calculated internally every time the value of Path changes

If we were doing this in C# we could do both of those things easily on the class with custom getter and setters and an internal property. PowerShell classes however do not have getters and setters so what we are left with is the Add-Member method for achieving that. I found doing it in the constructor to be pretty effective per [this stackoverflow suggestion via Rob Traynere](https://stackoverflow.com/questions/40977472/is-it-possible-to-override-the-getter-setter-functions-in-a-powershell-5-class).

The results are as displayed above with the added bonus of being able to manipulate the paths directly if needed, and the RegEx is auto-updated. Also we provide a warning message if the user attempts to set the value of RegEx directly:

![image](https://user-images.githubusercontent.com/6989492/56081680-118bac80-5dde-11e9-851a-2a706f1f8faa.png)

Reply:
Got around to reading this! I agree and think that's the right approach.